crates/matrix-sdk/src/widget/machine/mod.rs

```diff
@@ -306,16 +310,21 @@ impl WidgetMachine {
                 let request = ReadMessageLikeEventRequest { event_type, limit };
                 let (request, action) = self.send_matrix_driver_request(request);
                 request.then(|result, machine| {
-                    let response = result.and_then(|mut events| {
-                        let CapabilitiesState::Negotiated(capabilities) = &machine.capabilities
-                        else {
-                            let err = "Received read event request before capabilities negotiation";
-                            return Err(err.into());
-                        };
-
-                        events.retain(|e| capabilities.raw_event_matches_read_filter(e));
-                        Ok(ReadEventResponse { events })
-                    });
+                    let response = match (result, &machine.capabilities) {
+                        (Ok(mut events), CapabilitiesState::Negotiated(capabilities)) => {
+                            events.retain(|e| capabilities.raw_event_matches_read_filter(e));
+                            Ok(ReadEventResponse { events })
+                        }
+                        (Ok(_), CapabilitiesState::Unset) => {
+                            Err("Received read event request before capabilities negotiation"
+                                .into())
+                        }
+                        (Ok(_), CapabilitiesState::Negotiating) => {
+                            Err("Received read event request while capabilities were negotiating"
+                                .into())
+                        }
+                        (Err(e), _) => Err((&e).into()),
+                    };
```

Comment:
This is messier like this. Can we keep the `and_then`, and after it call `map_err` to convert the error type if needs be, instead?

Reply:
Are you sure with the custom error message dependent on the capability state this would become sth like this:
```
let CapabilitiesState::Negotiated(capabilities) = &machine.capabilities
else {
    match &machine.capabilities{
        CapabilitiesState::Unset => return  Err(FromWidgetErrorResponse::from_string(
            "Received read event request before capabilities negotiation",
        )),
        CapabilitiesState::Negotiating => return  Err(FromWidgetErrorResponse::from_string(
            "Received read event request while capabilities were negotiating",
        )),
        CapabilitiesState::Negotiated(capabilities) => todo!(),
    }
};
```